controller/controller_server/csi_controller_server.py

```diff
@@ -153,70 +156,64 @@ def CreateVolume(self, request, context):  # pylint: disable=too-many-branches
+            self._rollback_create_volume_from_source(array_mediator, volume.id)
             raise ex
-        return vol
 
     @retry(Exception, tries=5, delay=1)
-    def _rollback_create_volume_from_snapshot(self, array_mediator, vol_id):
-        logger.debug("Rollback copy volume from snapshot. Deleting volume {0}".format(vol_id))
+    def _rollback_create_volume_from_source(self, array_mediator, vol_id):
+        logger.debug("Rollback copy volume from source. Deleting volume {0}".format(vol_id))
         array_mediator.delete_volume(vol_id)
 
-    def _handle_existing_vol_src_snap(self, volume, src_snapshot_id, context):
+    def _handle_existing_volume_source(self, volume, source_object_id, source_type, context):
         """
         Args:
-            volume             : volume fetched or created in CreateVolume
-            src_snapshot_id    : id of snapshot we should copy to vol or None if volume should not be copied
-            context            : CreateVolume response context
+            volume              : volume fetched or created in CreateVolume
+            source_object_id       : id of object we should copy to vol or None if volume should not be copied
+            source_type:        : the object type of the source - volume or snapshot
+            context             : CreateVolume response context
         Returns:
-            If volume exists and is a copy of specified snapshot - set context status to OK
+            If volume exists and is a copy of specified object - set context status to OK
             and return CreateVolumeResponse.
             If volume is a copy of another source - set context status to INTERNAL and return CreateVolumeResponse.
             In any other case return None.
         """
-        vol_name = volume.volume_name
-        vol_copy_src_object_id = volume.copy_src_object_id
-        if not src_snapshot_id or not vol_copy_src_object_id:
+        volume_name = volume.name
+        volume_copy_source_object_id = volume.copy_src_object_id
+        if not source_object_id or not volume_copy_source_object_id:
             return None
-        if vol_copy_src_object_id == src_snapshot_id:
+        if volume_copy_source_object_id == source_object_id:
             logger.debug(
-                "Volume {0} exists and it is a copy of snapshot {1}.".format(vol_name, src_snapshot_id))
+                "Volume {0} exists and it is a copy of {1} {2}.".format(volume_name, source_type, source_object_id))
             context.set_code(grpc.StatusCode.OK)
-            return utils.generate_csi_create_volume_response(volume)
+            return utils.generate_csi_create_volume_response(volume, source_type)
         else:
             logger.debug(
-                "Volume {0} exists but it is not a copy of snapshot {1}.".format(vol_name, src_snapshot_id))
+                "Volume {0} exists but it is not a copy of object {1}.".format(volume_name, source_object_id))
```

Comment:
why not use `{source_type}` instead of `object`?

Reply:
done